### ISSUE TITLE
' System.Void ' error

### ISSUE THREAD

**Issue by sabmah** [agileobjects/AgileMapper#63]
Converting From

```
public class Location
{
  public int Id { get; set; } 
  public Place Place { get; set; }
}
public class Place
{
  public IEnumerable<PlaceLabel> PlaceLabels { get; set; }
  public IEnumerable<Location> Locations { get; set; }
}

public class PlaceLabel
{
        public virtual Label Label { get; set; }
        public virtual Place Place { get; set; }
}
```

To my Dto model 

```
public class Location
{
  public int Id { get; set; } 
  public Place Place { get; set; }
}
public class Place
{
  public IEnumerable<Label> Labels {get;set;}
  public IEnumerable<Location> Locations { get; set; }
}
public class Label
{
        public int Id { get; set; }
        public string Name { get; set; }
}
```

This is my plan on startup

```
AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Location>()
                .ToANew<Dto.Location>();

            AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Place>()
                .ToANew<Dto.Place>(
                cfg => cfg.Map(e => e.Source.PlaceLabels.Select(x => x.Label)).To(t => t.Labels)
                );
```

```
AgileObjects.AgileMapper.Mapper.Map<IEnumerable<Location>>(data).ToANew<IEnumerable<Dto.Location>>();
```

```
System.ArgumentException was unhandled by user code
  Message=Expression of type 'System.Void' cannot be used for parameter of type 'Dto.Label' of method 'Void Add(Dto.Label)'
  Source=System.Core
StackTrace:
       at System.Linq.Expressions.Expression.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arg, ParameterInfo pi)
       at System.Linq.Expressions.Expression.ValidateArgumentTypes(MethodBase method, ExpressionType nodeKind, ReadOnlyCollection`1& arguments)
       at System.Linq.Expressions.Expression.Call(Expression instance, MethodInfo method, IEnumerable`1 arguments)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.GetTargetMethodCall(String methodName, Expression argument)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.<AddNewItemsToTargetVariable>b__70_0(IPopulationLoopData ld, IObjectMappingData md)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.PopulationLoopDataExtensions.BuildPopulationLoop[TLoopData](TLoopData loopData, EnumerablePopulationBuilder builder, IObjectMappingData mappingData, Func`3 elementPopulationFactory)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.BuildPopulationLoop(Func`3 elementPopulationFactory, IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationBuilder.AddNewItemsToTargetVariable(IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.CopySourceEnumerablePopulationStrategy.GetEnumerablePopulation(EnumerablePopulationBuilder builder, IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerablePopulationStrategyBase.GetPopulation(IObjectMappingData enumerableMappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.Enumerables.EnumerableMappingExpressionFactory.<GetObjectPopulation>d__3.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.get_Mapper()
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IObjectMappingData childMappingData, MappingValues mappingValues, Int32 dataSourceIndex, ObjectMapperData declaredTypeMapperData)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IQualifiedMember sourceMember, Expression sourceMemberAccess, Int32 dataSourceIndex, IChildMemberMappingData childMappingData)
       at AgileObjects.AgileMapper.DataSources.EnumerableMappingDataSource.GetMapping(IDataSource sourceEnumerableDataSource, Int32 dataSourceIndex, IChildMemberMappingData enumerableMappingData)
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFindContext.GetFinalDataSource(IDataSource foundDataSource, IChildMemberMappingData mappingData)
       at AgileObjects.AgileMapper.DataSources.Finders.SourceMemberDataSourceFinder.GetSourceMemberDataSourceOrNull(DataSourceFindContext context)
       at AgileObjects.AgileMapper.DataSources.Finders.SourceMemberDataSourceFinder.<FindFor>d__0.MoveNext()
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.<EnumerateDataSources>d__2.MoveNext()
       at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
       at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.FindFor(IChildMemberMappingData childMappingData)
       at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.Create(QualifiedMember targetMember, IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.<>c__DisplayClass3_0.<Create>b__0(QualifiedMember tm)
       at System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource x)
       at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.<GetPopulationsAndCallbacks>d__3.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.<GetPopulation>d__0.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.get_Mapper()
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetInlineMappingBlock(IObjectMappingData mappingData, MappingValues mappingValues, Expression createMappingDataCall)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IObjectMappingData childMappingData, MappingValues mappingValues, Int32 dataSourceIndex, ObjectMapperData declaredTypeMapperData)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingFactory.GetChildMapping(IQualifiedMember sourceMember, Expression sourceMemberAccess, Int32 dataSourceIndex, IChildMemberMappingData childMappingData)
       at AgileObjects.AgileMapper.DataSources.ComplexTypeMappingDataSource.GetMapping(IDataSource complexTypeDataSource, Int32 dataSourceIndex, IChildMemberMappingData complexTypeMappingData)
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFindContext.GetFinalDataSource(IDataSource foundDataSource, IChildMemberMappingData mappingData)
       at AgileObjects.AgileMapper.DataSources.Finders.SourceMemberDataSourceFinder.GetSourceMemberDataSourceOrNull(DataSourceFindContext context)
       at AgileObjects.AgileMapper.DataSources.Finders.SourceMemberDataSourceFinder.<FindFor>d__0.MoveNext()
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.<EnumerateDataSources>d__2.MoveNext()
       at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
       at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
       at AgileObjects.AgileMapper.DataSources.Finders.DataSourceFinder.FindFor(IChildMemberMappingData childMappingData)
       at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.Create(QualifiedMember targetMember, IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.Members.Population.MemberPopulatorFactory.<>c__DisplayClass3_0.<Create>b__0(QualifiedMember tm)
       at System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource x)
       at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.<GetPopulationsAndCallbacks>d__3.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at AgileObjects.AgileMapper.ObjectPopulation.ComplexTypes.PopulationExpressionFactoryBase.<GetPopulation>d__0.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
       at AgileObjects.AgileMapper.ObjectPopulation.MappingExpressionFactoryBase.Create(IObjectMappingData mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.Create[TSource,TTarget](ObjectMappingData`2 mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.get_Mapper()
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.<>c__7`2.<GetOrCreateRoot>b__7_0(ObjectMapperKeyBase key)
       at AgileObjects.AgileMapper.Caching.ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapperFactory.GetOrCreateRoot[TSource,TTarget](ObjectMappingData`2 mappingData)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2..ctor(TSource source, TTarget target, Nullable`1 enumerableIndex, ObjectMapperKeyBase mapperKey, IMappingContext mappingContext, IObjectMappingData declaredTypeMappingData, IObjectMappingData parent)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, ObjectMapperKeyBase mapperKey, IMappingContext mappingContext)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](TSource source, TTarget target, IMappingContext mappingContext)
       at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingDataFactory.ForRootFixedTypes[TSource,TTarget](IMappingContext mappingContext)
       at AgileObjects.AgileMapper.Api.PlanTargetSelector`1.GetMappingPlan[TTarget](IMappingContext planContext, Func`2 mappingDataFactory, ICollection`1 configurations)
       at AgileObjects.AgileMapper.Api.PlanTargetSelector`1.GetMappingPlan[TTarget](MappingRuleSet ruleSet, ICollection`1 configurations)
       at AgileObjects.AgileMapper.Api.PlanTargetSelector`1.ToANew[TResult](Expression`1[] configurations)
```

**Comment by SteveWilkes** [comment]
Hi!

The plans cached for `Location` -> `Dto.Location` and `Place` -> `Dto.Place` are not used when you map from `IEnumerable<Location>` -> `IEnumerable<Dto.Location>`. Please try replacing your plan on start-up with:

```C#
AgileObjects.AgileMapper.Mapper
    .GetPlanFor<IEnumerable<Location>>()
    .ToANew<IEnumerable<Dto.Location>>(cfg => cfg
        .WhenMapping
        .From<Place>()
        .To<Dto.Place>()
        .Map((p, dto) => p.PlaceLabels)
        .To(dto => dto.Labels));
```

If that doesn't help, please could you include the class definitions for `PlaceLabel` and `Dto.Label`.

Thanks!

**Comment by sabmah** [comment]
@SteveWilkes  I updated my issue with ```PlaceLabel``` and ```Dto.Label``` definitions. 
Adding this plan didn't help:
```
AgileObjects.AgileMapper.Mapper
    .GetPlanFor<IEnumerable<Location>>()
    .ToANew<IEnumerable<Dto.Location>>(cfg => cfg
        .WhenMapping
        .From<Place>()
        .To<Dto.Place>()
        .Map((p, dto) => p.PlaceLabels)
        .To(dto => dto.Labels));

AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Location>()
                .ToANew<Dto.Location>();

AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Place>()
                .ToANew<Dto.Place>(
                cfg => cfg.Map(e => e.Source.PlaceLabels.Select(x => x.Label)).To(t => t.Labels)
                );
```

**Comment by SteveWilkes** [comment]
Thanks very much - I can see what you're trying to do now. I'll sort it out this week.

**Comment by SteveWilkes** [comment]
Hello again!

I've added the following test to my copy of the latest code:

```C#
Mapper
    .GetPlanFor<IEnumerable<Location>>()
    .ToANew<IEnumerable<Dto.Location>>(cfg => cfg
        .WhenMapping
        .From<Place>()
        .To<Dto.Place>()
        .Map((p, dto) => p.PlaceLabels.Select(pl => pl.Label))
        .To(dto => dto.Labels));

var location = new Location { Id = 1 };
var place = new Place();
var label = new Label { Id = 1, Name = "Mr Label" };
var placeLabel = new PlaceLabel { Place = place, Label = label };

location.Place = place;
place.PlaceLabels = new[] { placeLabel };
place.Locations = new[] { location };

var locationDtos = Mapper
    .Map<IEnumerable<Location>>(new[] { location })
    .ToANew<IEnumerable<Dto.Location>>();
```

...and the mapping works as you'd expect. The 'void' error is thrown when the `Location` -> `Dto.Location` mapping is cached as well.

As with the other recursion issue you've found, there is a bug here which I'll fix, but does it behave if you remove the `Location` -> `Dto.Location`  and `Place` -> `Dto.Place` caching calls?

**Comment by sabmah** [comment]
Thanks that worked as I excepted. 
**Side Note:** I already had Automapper implemented and wanted to see if AgileMapper is faster. But my analysis showed that the speed negligible. The example above returns 153 Json Items. Both Automapper and AgileMapper took 2.35s (average)

**Comment by SteveWilkes** [comment]
Ahh, I see - thanks for giving it a look :)

In a side note of my own - it looks like you're mapping from entities (virtual properties in `PlaceLabel`) - have you tried using [query projection](https://github.com/agileobjects/AgileMapper/wiki/Query-Projection)? AgileMapper [recognises](https://github.com/agileobjects/AgileMapper/wiki/Join-Entities) relationships like `Place` -> `PlaceLabel` -> `Label`, so shouldn't need that to be configured.

You'd use it like this:

```C#
var locationDtos = await sabmahDbContext
    .Locations
    .Project().To<Dto.Location>()
    .ToListAsync();
```

**Comment by SteveWilkes** [comment]
The underlying LocationRepository (assuming it's using Entity Framework) would use the projection mapping, yeah - it's faster because it modifies the query EF executes to only return the data from the database you need to create your DTOs, and you never materialize any entity objects, just the DTOs themselves. Architecturally speaking repositories usually surface entities rather than DTOs, though, so you might be on the best track if that's the way your application is structured. There are [other ways](https://lostechies.com/jimmybogard/2012/10/08/favor-query-objects-over-repositories/), though :)

AutoMapper does projection as well, btw - AgileMapper just does it more flexibly with less need for configuration.

**Comment by SteveWilkes** [comment]
Fixed in the latest code, release to follow.

**Comment by SteveWilkes** [comment]
Fixed in [v0.23.4](https://www.nuget.org/packages/AgileObjects.AgileMapper). Thanks again for the feedback! :)

**Comment by sabmah** [comment]
@SteveWilkes Thanks :)
